### ISSUE TITLE
Status bar does not support custom error types

### ISSUE THREAD

**Issue by kaste** [SublimeLinter/SublimeLinter#1726]
For example, it will print `(ok)` if there are only custom errors; otherwise it just shows (and counts) warnings and errors.

I suggest simplifying this one: We show just `(n)` if we have also custom types in the set of errors. For backwards compat we show `(n|n)` if we only count errors and warnings. We only show `(ok)` if the we really have no reported problems.

**Comment by braver** [comment]
Well we accepted the custom error types because they would be simple enough to support and these bugs were expected.

One way or the other, we need to introduce a new format here. The current format is customizable and always has both errors and warnings. There is no way to simplify based on that. 

We can introduce a new format that just bundles them all together, seriously reducing the usefulness of the statusbar information. Or we introduce a new format that splits them all out:

`(w|e|o)` (`o` for other)

We'd then only use that format when "other" error types are in the set.

**Comment by kaste** [comment]
I think the setting and even having a customizable format here is a typical bike-shedding option. In my opinion you should *always* avoid such options for the users, the app makers have to decide. Literally, just having that here makes the feature more complex. 

My initial reaction was to just do `(n)` and be done with it because in the end the file is `(ok)` or not. It is green or red. It's like a CI or build step. The first glance "icon" is "yes" or "no" then the user decides to look at the details, here usually doing `ctrl+k+a` for the panel, or clicking a link, here you can also click on the tiny panel menu (leftmost in the status bar; I actually do this not that seldom if I'm already on the mouse). 

I thought going with `(1|2|4)` is inherently a bad design. It is like the opposite of *design* in the sense that we design the UI in *how* we present the (usually to fine-grained) data; and instead we fall back to "here is all we have".

My current file shows

![image](https://user-images.githubusercontent.com/8558/78548630-63082500-7801-11ea-97e8-68a4d2457631.png)

Is that differentiation really useful?  I think just saying I have 4 annotations and 2 mypy errors is enough.  If I want the details I know where I get them. As a user, I also *know* mypy will fail the build, the annotations will not.

There are also a lot of linter only emitting one kind. "annotations" only emits warnings; "mypy" only emits "errors" and "notes"; "hlint" emits "warnings" and "suggestions" (for everything that strictly typechecks).

If we need to stick out "errors" then "others" and "warnings" can be merged (summed) (maybe); that would mean "errors" (like syntax errors) do not even compile; all "other" stuff like "warnings" will.  (If we apply these semantics to the results, an error *count* is still probably over-detailed: regardless of having one or two errors in the file, having an error would be a *red* flag, hence we could reduce it visually to a clear "nope! this will not work, please open the panel"-sign)

**Comment by braver** [comment]
> In my opinion you should always avoid such options for the users, the app makers have to decide. 

Sure I agree in the general sense. We should not have an option if we can agree on something that's better for everyone. But sometimes we can't. I can never remember if it's (errors|warnings) or (warnings|errors), so I have mine set up a bit more verbose (e:3|w:0). And now that we have the setting (and I believe a setting is warranted here), we cannot back track on that. Or we decide that my template is the best and we force that on everyone, but the default is not good enough for me. 

> Is that differentiation really useful?

Yes. In my team we set it up so that errors break the build and warnings don't. I cannot work with a general number that lumps everything together and I don't think that is a very weird setup. Why have any differentiation at all if the UI just throws everything into a big pile? Also, if you remove the differentiation here, it is then impossible for the user to find out the differentiated counts. 

> because in the end the file is (ok) or not. It is green or red.

It may be for you, but it is very common to allow warnings (and especially "note" or "information" level output), but not allow errors. 

I didn't really like adding the "other" category because it adds exactly this complication. I can *maybe* get behind saying "other" is a differentiation of "warning", but that's just such a random choice that I don't want to go there unless we must. Either we treat the "other" type fully, or we back track on the decision to support it in the first place. But let's not make some random assumptions about how thousands of users are using SL and how they treat the different error types, especially because between the two of us we already have different needs and we probably don't even agree on what the status bar template should be.

**Comment by braver** [comment]
tldr;

So, while I get why you want to have one lump sum in the status bar, I don't think you can get away with forcing that on all users. And then maybe you need to have a setting here 😄 It add complexity, but it's not like this adds a thousand lines of code.

**Comment by kaste** [comment]
Hm. It's actually because I never know which is the error and which is the warnings count I had the idea of just summing maybe. 

Generally, custom types are great. There is no need to reduce the data set because most views can print the detailed information just fine. (E.g. the panel did this from the beginning.)  In fact, these custom types were mapped by the plugins before which is even harder to get "right".  ("mypy" just transformed notes to warnings etc.)  The status bar is the only place were we aggressively aggregate.

Agg Aggregate is kind of lumps it altogether.  We have limited space so we cannot show detailed information.  Basically we have the default `(1|2)` because  we can only use ASCII and don't have any space.  And the default condenses the avail information in such a way that we both don't even know what the numbers are. So they might be room to improve.

**Comment by braver** [comment]
Well that makes a fair point and maybe the status bar just needs to aggregate for different reasons than other views. In that sense I am warming up to the idea of splitting into “errors” and “not errors”, ie. aggregate all custom error types into “warnings (and other stuff)”.

**Comment by kaste** [comment]
That's a pretty shitty "template" system here.

https://github.com/SublimeLinter/SublimeLinter/blob/046ec5f4dfb0145573fcb1addad5bcebe3155898/active_linters_view.py#L58-L58

t.i. we unpack a tuple and so everything we change is a breaking change. If we did `tpl.format(**we_count)` we sure had the same complexity but users could flip error with warnings, and we could just invent new keys like "total" etc. without breaking anyone.

**Comment by braver** [comment]
I noticed that... it’s kinda lame. 🙄 Maybe we can break here if we introduce something better. Or do it in a compatible way.
